cirq/qis/states.py

```diff
@@ -529,6 +534,45 @@ def to_valid_density_matrix(
     return np.outer(state_vector, np.conj(state_vector))
 
 
+def validate_density_matrix(
+        density_matrix: np.ndarray,
+        *,  # Force keyword arguments
+        qid_shape: Tuple[int, ...],
+        dtype: Optional[Type[np.number]] = None,
+        atol: float = 1e-7) -> None:
+    """Checks that the given density matrix is valid.
+
+    Args:
+        density_matrix: The density matrix to validate.
+        qid_shape: The expected qid shape.
+        dtype: The expected dtype.
+        atol: Absolute numerical tolerance.
+
+    Raises:
+        ValueError: The density matrix does not have the correct dtype.
+        ValueError: The density matrix is not square and of size 2**num_qubits.
+        ValueError: The density matrix is not Hermitian.
+        ValueError: The density matrix does not have trace 1.
+        ValueError: The density matrix is not positive semidefinite.
+    """
+    if dtype and density_matrix.dtype != dtype:
+        raise ValueError('Density matrix had dtype {} but expected {}'.format(
+            density_matrix.dtype, dtype))
+    if density_matrix.shape != (np.prod(qid_shape, dtype=int),) * 2:
+        raise ValueError(
+            'Density matrix was not square and of size 2**num_qubits, '
```

Comment:
Oh I see this is because this pice was extracted from the other function. Yeah we should definitely clean it up as `validate_denistiy_matrix` is a top level function now.

Reply:
I've improved the error messages.